wicker/core/datasets.py

```diff
@@ -1,30 +1,103 @@
 from wicker.core.shuffle import ShuffleWorker
-from wicker.core.storage import S3DataStorage, S3PathFactory
+from wicker.core.storage import (
+    AbstractDataStorage,
+    FileSystemDataStorage,
+    S3DataStorage,
+    S3PathFactory,
+    WickerPathFactory,
+)
 from wicker.schema import dataloading, serialization
 from wicker.schema.schema import DatasetSchema
 
 # How long to wait before timing out on filelocks in seconds
 FILE_LOCK_TIMEOUT_SECONDS = 300
 
+logger = logging.getLogger(__name__)
+
 
 class AbstractDataset(abc.ABC):
     """Interface for a Map-style (non-streaming) Dataset"""
 
+    def __init__(
+        self,
+        dataset_name: str,
+        dataset_partition_name: str,
+        dataset_version: str,
+        path_factory: WickerPathFactory,
+        pa_filesystem: pafs.FileSystem,
+        storage: AbstractDataStorage,
+        columns_to_load: Optional[List[str]] = None,
+        filelock_timeout_seconds: int = FILE_LOCK_TIMEOUT_SECONDS,
+        local_cache_path_prefix: Optional[str] = os.getenv("TMPDIR", "/tmp"),
+        treat_objects_as_bytes: bool = False,
+    ) -> None:
+        """Init an AbstractDataset object.
+
+        :param dataset_name: name of the dataset
+        :param dataset_partition_name: partition name
+        :param dataset_version: version of the dataset
+        :param path_factory: WickerPathFactory for pulling consistent paths.
+        :param pa_filesystem: Pyarrow filesystem for reading the parquet files and tables.
+        :param storage: AbstractDataStorage for data access.
+        :param columns_to_load: list of columns to load, defaults to None which loads all columns
+        :param filelock_timeout_seconds: number of seconds after which to timeout on waiting for downloads,
+            defaults to FILE_LOCK_TIMEOUT_SECONDS
+        :param local_cache_path_prefix: Path to local cache path, if None don't create cache
+        :param treat_objects_as_bytes: If set, don't try to decode ObjectFields and keep them as binary data.
+        """
+        super().__init__()
+        self._arrow_table: Optional[pyarrow.Table] = None
+        self._columns_to_load = columns_to_load
+        self._filelock_timeout_seconds = filelock_timeout_seconds
+        self._local_cache_path_prefix = local_cache_path_prefix
+        self._pa_filesystem = pa_filesystem
+        self._path_factory = path_factory
+        self._storage = storage
+        self._treat_objects_as_bytes = treat_objects_as_bytes
+
+        # if we have a cache prefix create a cache
+        if local_cache_path_prefix is not None:
+            logging.info(f"Cache passed at path - {local_cache_path_prefix}, creating read through cache on top.")
+            # weird mypy problem, can't define baseclass and have it pick up the child correctly
+            self._column_bytes_file_reader: Union[ColumnBytesFileReader, ColumnBytesFileCache] = ColumnBytesFileCache(
```

Comment:
Lol sorry about having to add the type annotation. However, the if-else code paths are a lot easier to read to me now though.